### PR TITLE
Add direct_connect_gateway_id to virtual_interface

### PR DESCRIPTION
**Pull request by mtraynham**
##### SUMMARY
This adds `direct_connect_gateway_id` to `aws_direct_connect_virtual_interface`.
This field is only applicable in private VIF cases (`public=False`) and is
mutually exclusive to `virtual_gateway_id`.

The initial PR for the existing ansible/ansible repo was here, https://github.com/ansible/ansible/pull/48711.  It is very similar, but lacked return documentation and had a bug that allowed `direct_connect_gateway_id` for `public=True` VIFs, which is not supported by AWS.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
aws_direct_connect_virtual_interface

### CODE REVIEW

**Comment by mtraynham** [comment]
@jillr @tremble , I've added unit tests for `aws_direct_connect_virtual_interface`.  

I had to comment out two tests which actually performed the boto3 execution for creating VIFs.  For some reason placeboify is not playing the second fixture, `directconnect.DescribeVirtualInterfaces_2.json` in both cases, and I'm not really sure why.  Maybe you two might have some feedback?

File: tests/unit/modules/test_aws_direct_connect_virtual_interface.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                        bgp_asn=123,
+                        authentication_key="aaaa",
+                        customer_address="169.254.0.1/30",
+                        amazon_address="169.254.0.2/30",
+                        address_type="ipv4",
+                        cidr=["10.88.0.0/30"],
+                        virtual_gateway_id="xxxx",
+                        direct_connect_gateway_id="yyyy")
+    vi = aws_direct_connect_virtual_interface.assemble_params_for_creating_vi(module.params)
+    print(vi)
```
Comment (jillr) [inline tests/unit/modules/test_aws_direct_connect_virtual_interface.py:122]: Could you please remove print statements from the tests?

**Comment by mtraynham** [comment]
@jillr Thanks for taking a look at this!  

I'm looking at the LAG tests as you mentioned, but I'm a bit unsure how that fits into these tests and maybe it's just a misunderstanding on my part.  The LAG tests seem to be setting up the dependencies for a single pytest to run multiple times with different inputs.  Rather, for the tests I wrote, I was looking to have a single boto3 function return a different response on it's second invocation.

For example, I've pasted one of the tests I have commented out is below.  The `aws_direct_connect_virtual_interface.ensure_state` function should initially call `directconnect.DescribeVirtualInterfaces_1` and that returns an empty response, which intends to create a new VIF.  Then `ensure_state` would construct the VIF, post to AWS and finally calls `client.describeVirtualInterafaces` a second time to get the newly created VIF, e.g. from the placebo response `directconnect.DescribeVirtualInterfaces_2`.

If I'm way off here, feel free to let me know 😄 

```python
def test_create_public_vi(placeboify, maybe_sleep):
    client = placeboify.client("directconnect")
    module = FakeModule(state="present",
                        id_to_associate='dxcon-aaaaaaaa',
                        virtual_interface_id=None,
                        public=True,
                        name="aaaaaaaa",
                        vlan=1,
                        bgp_asn=123,
                        authentication_key="aaaa",
                        customer_address="169.254.0.1/30",
                        amazon_address="169.254.0.2/30",
                        address_type="ipv4",
                        cidr=["10.88.0.0/30"],
                        virtual_gateway_id="xxxx",
                        direct_connect_gateway_id="yyyy")
    changed, latest_state = aws_direct_connect_virtual_interface.ensure_state(client, module)
    assert changed is True
    assert latest_state is not None
```

**Comment by jillr** [comment]
@mtraynham I apologize for the delay in responding, you're right the LAG tests were unrelated.  I've had a chance to work through this a bit more and made some progress but am not yet sure exactly why that test is going wrong.  It looks like `client.describe_virtual_interfaces` is not being called the expected number of times for the code path it looks like that test should be following.  It seems like `find_unique_vi()` is not being called at ln446, leading to  `directconnect.DescribeVirtualInterfaces_1.json` is being read in to `vi_state()` which fails as there's no `vi['virtualInterfaces'][0]`.  Copying `DescribeVirtualInterfaces_2.json` to `DescribeVirtualInterfaces_1.json` so that there's at least a list element there for this then results in a failure on `assert changed is True` (did not have a chance to pin down why changed status is not getting set).

Next best steps would be to track down why the code path this test is following doesn't follow our expectations.  You may be able to use mocker.patch describe_virtual_interfaces to get a call count or set some side effects or print the stack to see where things are going.  I will try to dig into this further when I can.

**Comment by jillr** [comment]
Hi @mtraynham, wanted to share the 2.10 [roadmap](https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_2_10.html) dates with you.  Collections will have a freeze on new modules on 2020-08-18.  It would be great to get these tests in order so we could include the direct connect modules in that build possible.

**Comment by mtraynham** [comment]
Thanks for the update @jillr.  I added some print statements to help me figure out the issue.  Seems retrieving the direct connect occurs 3 times in those test suites and not 2.  Therefore placeboify was I guess replaying from the start, expecting 0 VIFs to exist.

**Pull request merged**